lightning/mydump/router.go

```diff
@@ -0,0 +1,365 @@
+package mydump
+
+import (
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/pingcap/tidb-lightning/lightning/config"
+
+	"github.com/pingcap/errors"
+
+	"github.com/pingcap/tidb-tools/pkg/filter"
+)
+
+type SourceType int
+
+const (
+	SourceTypeIgnore SourceType = iota
+	SourceTypeSchemaSchema
+	SourceTypeTableSchema
+	SourceTypeSQL
+	SourceTypeCSV
+)
+
+const (
+	SchemaSchema = "schema-schema"
+	TableSchema  = "table-schema"
+	TypeSQL      = "sql"
+	TypeCSV      = "csv"
+	TypeIgnore   = "ignore"
+)
+
+type Compression int
+
+const (
+	CompressionNone Compression = iota
+	CompressionGZ
+	CompressionLZ4
+	CompressionZStd
+	CompressionXZ
+)
+
+func parseSourceType(t string) SourceType {
+	switch strings.ToLower(strings.TrimSpace(t)) {
+	case SchemaSchema:
+		return SourceTypeSchemaSchema
+	case TableSchema:
+		return SourceTypeTableSchema
+	case TypeSQL:
+		return SourceTypeSQL
+	case TypeCSV:
+		return SourceTypeCSV
+	default:
+		return SourceTypeIgnore
+	}
+}
+
+func (s SourceType) String() string {
+	switch s {
+	case SourceTypeSchemaSchema:
+		return SchemaSchema
+	case SourceTypeTableSchema:
+		return TableSchema
+	case SourceTypeCSV:
+		return TypeCSV
+	case SourceTypeSQL:
+		return TypeSQL
+	default:
+		return TypeIgnore
+	}
+}
+
+func parseCompressionType(t string) Compression {
+	switch strings.ToLower(strings.TrimSpace(t)) {
+	case "gz":
+		return CompressionGZ
+	case "lz4":
+		return CompressionLZ4
+	case "zstd":
+		return CompressionZStd
+	case "xz":
+		return CompressionXZ
+	default:
+		return CompressionNone
+	}
+}
+
+var (
+	isAlphaNum = regexp.MustCompile(`^[A-Za-z0-9]+$`).MatchString
+	isAlpha    = regexp.MustCompile(`^[0-9]+$`).MatchString
+)
+
+var (
+	defaultFileRouteRules = []*config.FileRouteRule{
+		// ignore *-schema-view.sql,-schema-trigger.sql,-schema-post.sql files
+		{Pattern: `(?i).*(-schema-view|-schema-trigger|-schema-post)\.sql`, Type: "ignore"},
+		// db schema create file pattern, matches files like '{schema}-schema-create.sql'
+		{`(?i)^(?:[^/]*/)*([^/.]+)-schema-create\.sql`, "$1", "", SchemaSchema, "", ""},
+		// table schema create file pattern, matches files like '{schema}.{table}-schema.sql'
+		{`(?i)^(?:[^/]*/)*([^/.]+)\.(.*?)-schema\.sql`, "$1", "$2", TableSchema, "", ""},
+		// source file pattern, matches files like '{schema}.{table}.0001.{sql|csv}'
+		{`(?i)^(?:[^/]*/)*([^/.]+)\.(.*?)(\.[0-9]+)?\.(sql|csv)$`, "$1", "$2", "$4", "$3", ""},
+	}
+)
+
+// // RouteRule is a rule to route file path to target schema/table
+type FileRouter interface {
```

Comment:
why this is an interface, instead of a slice of `RegexRouteRule`

Reply:
I think there is no much difference to make it as a interface or an actual implement type. I prefer to use an interface type, because it's more clear for use to determine the methods this type provide.

Reply:
And I think maybe, we can Implement another `FileRouter` by just treat the pattern as file path, in this way, user needn't escape special chars in the file path, though I'm not sure how should we determine whether the input is an regex pattern or just a plain file path.

Reply:
i see. perhaps always use `pattern = ...` for patterns and introduce `path = ...` for literal paths.